### PR TITLE
Add GetAllKeyVersions

### PR DESCRIPTION
**Pull request by follitude**

### CODE REVIEW

File: librocksdb_sys/src/lib.rs [hunk 1]
```diff
@@ -976,6 +980,46 @@ extern "C" {
                                                       props: *mut DBTablePropertiesCollection,
                                                       errptr: *mut *mut c_char);
 
+    // Get All Key Versions
+    pub fn crocksdb_keyversions_create() -> *mut DBKeyVersions;
+
+    pub fn crocksdb_keyversions_destroy(kvs: *mut DBKeyVersions);
+
+    pub fn crocksdb_keyversions_iterator_create(kv: *mut DBKeyVersions)
```
Comment (siddontang) [inline librocksdb_sys/src/lib.rs]: why do we support iterator here? and benefit?

Comment (follitude) [inline librocksdb_sys/src/lib.rs]: Because we need to get a vector.

Comment (siddontang) [inline librocksdb_sys/src/lib.rs]: We know the DBKeyVersions and its count, why not using a function like crocksdb_keyversions_get(kvs: * DBKeyVersions, index: usize) to get the version directly?

Comment (follitude) [inline librocksdb_sys/src/lib.rs]: Actually, we don't know the count. We only give a begin_key and an end_key to the GetAllKeyVersions function. If we wanna get the size, we need to run the function two times.

Comment (siddontang) [inline librocksdb_sys/src/lib.rs]: why can't we know? you can return the count in crocksdb_get_all_key_versions.

Comment (siddontang) [inline librocksdb_sys/src/lib.rs]: btw, the return of crocksdb_get_all_key_versions in c is void, but in rust is size_t.

Comment (follitude) [inline librocksdb_sys/src/lib.rs]: Ok, I will fix it.

Comment (siddontang) [inline librocksdb_sys/src/lib.rs]: or you can add a get_count function like c_rocksdb_keyversions_count(kvs: *DBKeyVersions) to get count first.

File: librocksdb_sys/crocksdb/c.cc [hunk 2]
```diff
@@ -3299,8 +3314,79 @@ void crocksdb_get_properties_of_tables_in_range(
...
+}
+
+crocksdb_keyversion_t *crocksdb_keyversion_create() {
+  return new crocksdb_keyversion_t;
+}
+
+void crocksdb_keyversion_destroy(crocksdb_keyversion_t *kv) { delete kv; }
+
+void crocksdb_get_all_key_versions(crocksdb_t *db, const char *begin_key,
+                                   size_t begin_keylen, const char *end_key,
```
Comment (zhangjinpeng1987) [inline librocksdb_sys/crocksdb/c.cc]: s/begin_key/start_key
s/begin_keylen/start_key_len

File: librocksdb_sys/crocksdb/c.cc [hunk 3]
```diff
@@ -3299,8 +3314,79 @@ void crocksdb_get_properties_of_tables_in_range(
...
+                                   size_t end_keylen,
```
Comment (zhangjinpeng1987) [inline librocksdb_sys/crocksdb/c.cc]: s/end_keylen/end_key_len

**Comment by siddontang** [comment]
any update?

**Comment by siddontang** [comment]
LGTM

**Comment by siddontang** [comment]
PTAL @zhangjinpeng1987

File: src/rocksdb.rs [hunk 4]
```diff
@@ -306,6 +306,13 @@ impl<'a> Range<'a> {
     }
 }
 
+pub struct KeyVersion {
+    pub key: String,
+    pub value: String,
+    pub seq: u64,
+    pub key_type: c_int,
```
Comment (zhangjinpeng1987) [inline src/rocksdb.rs]: Please use `DBEntryType`

File: src/lib.rs [hunk 5]
```diff
@@ -30,6 +30,7 @@ mod table_properties;
 mod table_properties_collector;
 mod table_properties_collector_factory;
 
+
```
Comment (zhangjinpeng1987) [inline src/lib.rs]: remove blank line

Comment (follitude) [inline src/lib.rs]: fixed @zhangjinpeng1987

**Pull request merged**